lib/src/main/java/graphql/nadel/validation/NadelTypeValidation.kt

```diff
@@ -229,16 +232,30 @@ internal class NadelTypeValidation(
             } to errors
     }
 
-    private fun getPolymorphicHydrationUnions(service: Service): Set<GraphQLUnionType> {
+    private fun getHydrationUnions(service: Service): Set<GraphQLUnionType> {
         return service.definitionRegistry
             .definitions
             .asSequence()
-            .filterIsInstance<ObjectTypeDefinition>()
-            .flatMap { it.fieldDefinitions }
-            .filter { it.getDirectives(hydratedDirectiveDefinition.name).size > 1 }
-            .map { it.type.unwrapAll() }
-            .map { overallSchema.getType(it.name) }
-            .filterIsInstance<GraphQLUnionType>()
+            .filterIsInstance<UnionTypeDefinition>()
+            .filter { union ->
+                // Check that ALL fields that output the union are annotated with @hydrated
+                overallSchema.typeMap
+                    .values
+                    .asSequence()
+                    .filterIsInstance<GraphQLFieldsContainer>()
+                    .flatMap {
+                        it.fieldDefinitions
+                    }
+                    .filter {
+                        it.type.unwrapAll().name == union.name
+                    }
+                    .all(min = 1) {
```

Comment:
I'm saying here we need _at least_ one `@hydrated` field that uses this.

Maybe we could drop this requirement… but I kept it in to satisfy some tests regarding not allowing synthetic unions. (no hydrations referenced in that test).